File: android/src/main/kotlin/com/fluxpert/gma_all_mediations/GmaAllMediationsPlugin.kt
// Copyright 2024 FluXpert. All rights reserved.
//
// GmaAllMediationsPlugin.kt
//
// Native Android plugin for the gma_all_mediations Flutter package.
//
// Handles the "gma_all_mediations/chartboost_consent" MethodChannel.
//
// On Android, the Chartboost mediation adapter reads GDPR / CCPA consent
// signals automatically via the Google Mobile Ads mediation framework.
// No explicit call to Chartboost.setDataUseConsent() is required on this
// platform — the adapter propagates consent through the standard
// RequestConfiguration set on MobileAds.
//
// This plugin therefore acts as a no-op for the "applyConsent" method on
// Android, while still sending a log entry for observability.

package com.fluxpert.gma_all_mediations

import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * Flutter plugin that registers the Chartboost consent MethodChannel.
 *
 * On Android, Chartboost reads consent automatically via the mediation
 * adapter — no explicit SDK call is needed. The "applyConsent" handler
 * therefore succeeds silently so the Dart layer behaves identically on
 * both platforms.
 */
class GmaAllMediationsPlugin : FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel

  // ── FlutterPlugin ─────────────────────────────────────────────────────

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(
      binding.binaryMessenger,
      "gma_all_mediations/chartboost_consent"
    )
    channel.setMethodCallHandler(this)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  // ── MethodCallHandler ─────────────────────────────────────────────────

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "applyConsent" -> handleApplyConsent(call, result)
      else -> result.notImplemented()
    }
  }

  // ── Private helpers ───────────────────────────────────────────────────

  /**
   * Handles the "applyConsent" call from Dart.
   *
   * On Android, the Chartboost adapter inherits consent from the RequestConfiguration
   * already applied to MobileAds by [GmaAllMediations._startMobileAdsSdk()].
   * No additional Chartboost-specific call is required.
   *
   * @param call Contains "hasConsent" (Boolean) and "doNotSell" (Boolean) arguments.
   * @param result Flutter result callback — always succeeds on Android.
   */
  private fun handleApplyConsent(call: MethodCall, result: Result) {
    val hasConsent = call.argument<Boolean>("hasConsent") ?: false
    val doNotSell = call.argument<Boolean>("doNotSell") ?: false

    // Android: Chartboost reads consent via the mediation adapter automatically.
    // The explicit addDataUseConsent() call is an iOS-only requirement.
    Log.i(
      "GMA",
      "✅  Chartboost — Android adapter registered " +
      "(consent auto-managed). hasConsent=$hasConsent, doNotSell=$doNotSell"
    )

    result.success(null)
  }
}

File: ios/Classes/GmaAllMediationsPlugin.swift
// Copyright 2024 FluXpert. All rights reserved.
//
// GmaAllMediationsPlugin.swift
//
// Native iOS plugin for the gma_all_mediations Flutter package.
//
// Responsibilities:
//  • Registers the "gma_all_mediations/chartboost_consent" MethodChannel.
//  • Handles the "applyConsent" method to forward GDPR and CCPA signals
//    to the Chartboost SDK via Chartboost.addDataUseConsent().
//
// The Chartboost SDK is brought in transitively through the
// gma_mediation_chartboost pod (GoogleMobileAdsMediationChartboost ~> 9.x).
// No additional pod dependency is required.
//
// Why natively?
// The gma_mediation_chartboost Flutter package exposes an intentionally
// empty Dart class — Google's design delegated consent to native.
// This plugin closes that gap so Flutter consumers have zero AppDelegate work.

import Flutter
import UIKit

@objc protocol DynamicCHBGDPRDataUseConsent: NSObjectProtocol {
  @objc(initWithValue:)
  init(_ value: Int)
}

@objc protocol DynamicCHBCCPADataUseConsent: NSObjectProtocol {
  @objc(initWithValue:)
  init(_ value: Int)
}

@objc protocol DynamicChartboost: NSObjectProtocol {
  @objc(addDataUseConsent:)
  static func addDataUseConsent(_ consent: AnyObject)
}

/// Flutter plugin that bridges Chartboost consent from Dart to the
/// Chartboost iOS SDK.
///
/// Registered automatically via the Flutter plugin system. Consumers do
/// **not** need to add anything to their `AppDelegate`.
public class GmaAllMediationsPlugin: NSObject, FlutterPlugin {
  // ── FlutterPlugin registration ──────────────────────────────────────────

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = GmaAllMediationsPlugin()

    // Chartboost channel
    let chartboostChannel = FlutterMethodChannel(
      name: "gma_all_mediations/chartboost_consent",
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: chartboostChannel)

    // Meta (Facebook) channel
    let metaChannel = FlutterMethodChannel(
      name: "gma_all_mediations/meta_consent",
      binaryMessenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: metaChannel)
  }

  // ── MethodCall handler ──────────────────────────────────────────────────

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "applyConsent":
      applyChartboostConsent(call: call, result: result)
    case "setAdvertiserTrackingEnabled":
      applyMetaTracking(call: call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // ── Private helpers ─────────────────────────────────────────────────────

  /// Applies GDPR and CCPA consent flags to the Chartboost SDK.
  ///
  /// Arguments expected in `call.arguments`:
  /// - `"hasConsent"` (Bool) – `true` → behavioural ads; `false` → non-behavioural.
  /// - `"doNotSell"`  (Bool) – `true` → opt-out of sale (CCPA); `false` → opt-in.
  private func applyChartboostConsent(
    call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) {
    guard let args = call.arguments as? [String: Any],
          let hasConsent = args["hasConsent"] as? Bool,
          let doNotSell = args["doNotSell"] as? Bool
    else {
      result(
        FlutterError(
          code: "INVALID_ARGUMENTS",
          message: "applyConsent requires 'hasConsent' (Bool) and 'doNotSell' (Bool).",
          details: nil
        )
      )
      return
    }

    // Safely look up Chartboost classes via runtime reflection.
    // If the app does not link ChartboostSDK, this silently does nothing.
    guard let chartboostClass = NSClassFromString("Chartboost") as? DynamicChartboost.Type,
          let gdprClass = NSClassFromString("CHBGDPRDataUseConsent") as? DynamicCHBGDPRDataUseConsent.Type,
          let ccpaClass = NSClassFromString("CHBCCPADataUseConsent") as? DynamicCHBCCPADataUseConsent.Type
    else {
      result(nil)
      return
    }

    // ── GDPR ──────────────────────────────────────────────────────────────
    // 0 = non-behavioral, 1 = behavioral
    let gdprConsent = gdprClass.init(hasConsent ? 1 : 0)
    chartboostClass.addDataUseConsent(gdprConsent)

    // ── CCPA ──────────────────────────────────────────────────────────────
    // 0 = opt-in sale, 1 = opt-out sale
    let ccpaConsent = ccpaClass.init(doNotSell ? 1 : 0)
    chartboostClass.addDataUseConsent(ccpaConsent)

    result(nil) // Success
  }

  /// Forwards the ATT tracking authorization status to the Meta SDK.
  ///
  /// Arguments expected in `call.arguments`:
  /// - `"trackingEnabled"` (Bool) – `true` → authorized; `false` → not authorized.
  ///
  /// Uses Objective-C runtime reflection so that this plugin does not require
  /// a compile-time dependency on `FBAudienceNetwork`. If the app does not
  /// use the Meta adapter, this silently does nothing.
  private func applyMetaTracking(
    call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) {
    guard let args = call.arguments as? [String: Any],
          let trackingEnabled = args["trackingEnabled"] as? Bool
    else {
      result(
        FlutterError(
          code: "INVALID_ARGUMENTS",
          message: "setAdvertiserTrackingEnabled requires 'trackingEnabled' (Bool).",
          details: nil
        )
      )
      return
    }

    // Safely look up FBAdSettings without importing FBAudienceNetwork.
    if let fbAdSettingsClass: AnyClass = NSClassFromString("FBAdSettings") {
      let selector = NSSelectorFromString("setAdvertiserTrackingEnabled:")
      if fbAdSettingsClass.responds(to: selector),
         let targetClass = fbAdSettingsClass as? NSObjectProtocol {
        targetClass.perform(selector, with: NSNumber(value: trackingEnabled))
      }
    }

    result(nil)
  }
}
